stan-dev/rstan
# Make V8 stancjs in .OnLoad()

#### Summary:

This hopefully fixes https://github.com/stan-dev/rstan/issues/945 and https://github.com/rstudio/rstudio/issues/9282 by moving the stancjs setup logic to `.OnLoad()`. 

I know how busy @jeroen but Jeroen if you would have time to take a peek at this it would be very helpful. Essentially I think the problem is that v8 is a global object and calling `ctx$source(stanc_file)` was causing memory to continuously be used. Calling `gc()` wouldn't help with this since v8 is a global.

I also removed `allow_optimizations` from stanc and friends. Those are super experimental and can potentially produce incorrect C++ and are not ready to be exposed to users. My big concern here is that users are going to see "allow optimizations" and go "well of course yes" without actually knowing they are experimental / dangerous. We are also working right now to have optimization levels like 1,2,3 etc. so that interface would change anyway.

#### Intended Effect:

Fixes the above memory leaks

#### How to Verify:

I'm actually not sure how to verify this I was just taking an example model and doing 

```r
for (i in 1:500) {
  stanc_builder("../test_mod.stan")
}
```

and checking the memory usage via htop to see if it kept going up.

#### Side Effects:

#### Documentation:

#### Reviewer Suggestions: 

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Steve Bronder

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: GPLv3 (http://opensource.org/licenses/GPL-3.0)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)
